HDFEventOutputer.cc

```diff
@@ -0,0 +1,204 @@
+  }
+}
+
+HDFEventOutputer::HDFEventOutputer(std::string const& iFileName, unsigned int iNLanes, int iBatchSize, pds::Serialization iSerialization) : 
+  file_(hdf5::File::create(iFileName.c_str())),
+  maxBatchSize_{iBatchSize},
+  serializers_{std::size_t(iNLanes)},
+  serialization_{iSerialization},
+  serialTime_{std::chrono::microseconds::zero()},
+  parallelTime_{0}
+  {}
+
+HDFEventOutputer::~HDFEventOutputer() { }
+
+void HDFEventOutputer::setupForLane(unsigned int iLaneIndex, std::vector<DataProductRetriever> const& iDPs) {
+  auto& s = serializers_[iLaneIndex];
+  switch(serialization_) {
+  case pds::Serialization::kRoot:
+    {   s = SerializeStrategy::make<SerializeProxy<SerializerWrapper>>(); break; }
+  case pds::Serialization::kRootUnrolled:
+    {   s = SerializeStrategy::make<SerializeProxy<UnrolledSerializerWrapper>>(); break; }
+  }
+  s.reserve(iDPs.size());
+  offsetsAndBlob_.first.resize(iDPs.size()+1, 0);
+  for(auto const& dp: iDPs) {
+    s.emplace_back(dp.name(), dp.classType());
+  }
+  std::cout << "Setting up for Lanes\n"; 
+  if(iLaneIndex == 0) {writeFileHeader(s); }
+}
+
+void HDFEventOutputer::productReadyAsync(unsigned int iLaneIndex, DataProductRetriever const& iDataProduct, TaskHolder iCallback) const {
+  auto& laneSerializers = serializers_[iLaneIndex];
+  auto group = iCallback.group();
+  laneSerializers[iDataProduct.index()].doWorkAsync(*group, iDataProduct.address(), std::move(iCallback));
+}
+
+void HDFEventOutputer::outputAsync(unsigned int iLaneIndex, EventIdentifier const& iEventID, TaskHolder iCallback) const {
+  auto start = std::chrono::high_resolution_clock::now();
+  auto [offsets, buffer] = writeDataProductsToOutputBuffer(serializers_[iLaneIndex]);
+  queue_.push(*iCallback.group(), [this, iEventID, iLaneIndex, callback=std::move(iCallback), buffer = std::move(buffer), offsets = std::move(offsets)]() mutable {
+      auto start = std::chrono::high_resolution_clock::now();
+      const_cast<HDFEventOutputer*>(this)->output(iEventID, serializers_[iLaneIndex], std::move(buffer), std::move(offsets));
+        serialTime_ += std::chrono::duration_cast<decltype(serialTime_)>(std::chrono::high_resolution_clock::now() - start);
+      callback.doneWaiting();
+    });
+    auto time = std::chrono::duration_cast<std::chrono::microseconds>(std::chrono::high_resolution_clock::now() - start);
+    parallelTime_ += time.count();
+}
+
+void HDFEventOutputer::printSummary() const  {
+  std::cout <<"HDFEventOutputer\n  total serial time at end event: "<<serialTime_.count()<<"us\n"
+    "  total parallel time at end event: "<<parallelTime_.load()<<"us\n";
+  summarize_serializers(serializers_);
+}
+
+
+
+void 
+HDFEventOutputer::output(EventIdentifier const& iEventID, 
+                         SerializeStrategy const& iSerializers,
+                         std::vector<char> iBuffer,
+                         std::vector<uint32_t> iOffsets ) {
+  hdf5::Group gid = hdf5::Group::open(file_, "Lumi");   
```

Comment:
Could this be done in the constructor and the group be cached in member data? That way we avoid a per event call. Or is the `open` really an `openTransaction` and has to be done each time?

Reply:
it can be with the current design for sure, but when in future we do multiple lumi/file design we will need to see what changes will be needed then. and I will make group name as command line/configuration option. so that it is not hard coded as "lumi" .